File: xabber/src/main/java/org/jivesoftware/smackx/mam/provider/MamFinIQProvider.java
package org.jivesoftware.smackx.mam.provider;

import org.jivesoftware.smack.provider.IQProvider;
import org.jivesoftware.smack.util.ParserUtils;
import org.jivesoftware.smackx.mam.packet.MamFinIQ;
import org.jivesoftware.smackx.rsm.packet.RSMSet;
import org.jivesoftware.smackx.rsm.provider.RSMSetProvider;
import org.xmlpull.v1.XmlPullParser;

public class MamFinIQProvider extends IQProvider<MamFinIQ> {
    @Override
    public MamFinIQ parse(XmlPullParser parser, int initialDepth) throws Exception {
        boolean done = false;

        MamFinIQ mamFinIQ = new MamFinIQ();

        mamFinIQ.setComplete(ParserUtils.getBooleanAttribute(parser, "complete"));

        RSMSet rsmSet = null;
        while (!done) {
            int eventType = parser.next();
            if (eventType == XmlPullParser.START_TAG) {
                if (parser.getName().equals("set")) {
                    rsmSet = new RSMSetProvider().parse(parser);
                }
            } else {
                if (eventType == XmlPullParser.END_TAG && parser.getDepth() == initialDepth) {
                    done = true;
                }
            }
        }

        mamFinIQ.setRsmSet(rsmSet);

        return mamFinIQ;
    }
}

File: xabber/src/main/java/com/xabber/xmpp/mam/MamFinIqProvider.kt
package com.xabber.xmpp.mam

import org.jivesoftware.smack.provider.IQProvider
import org.jivesoftware.smack.util.ParserUtils
import org.jivesoftware.smackx.rsm.packet.RSMSet
import org.jivesoftware.smackx.rsm.provider.RSMSetProvider
import org.xmlpull.v1.XmlPullParser

class MamFinIqProvider : IQProvider<MamFinIQ>() {

    override fun parse(parser: XmlPullParser, initialDepth: Int): MamFinIQ {
        val queryId = parser.getAttributeValue("", MamFinIQ.QUERY_ID_ATTRIBUTE)
        val complete = ParserUtils.getBooleanAttribute(parser, MamFinIQ.COMPLETE_ATTRIBUTE, false)
        var rsmSet: RSMSet? = null

        outerloop@ while (true) {
            when (parser.next()) {
                XmlPullParser.START_TAG -> {
                    if (parser.name == RSMSet.ELEMENT) {
                        rsmSet = RSMSetProvider.INSTANCE.parse(parser)
                    }
                }
                XmlPullParser.END_TAG -> if (parser.depth == initialDepth) {
                    break@outerloop
                }
            }
        }

        return MamFinIQ(
            rsmSet = rsmSet!!,
            isComplete = complete,
            queryId = queryId
        )
    }

}